docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+* Either shoot owner creates shoot from scratch or needs to align with Gardener operator who has to move the shoot to a proper seed first via control plane migration (editing the `shoots/binding` resource).
+* An automated control plane migration is deliberately not performed as it involves a considerable downtime and the feature itself is not stable by the time this GEP was written.
+
+## Setting up a Seed for HA
+
+As mentioned in [High-Availablity](#high-availablity), certain aspects need to be considered for a seed cluster to host HA shoots. The following sections explain the requirements for a seed cluster to host a single or multi zonal HA shoot cluster.
+
+### Hosting a HA shoot control plane with `node` failure tolerance 
+
+To host an HA shoot control plane within a single zone, it should be ensured that each worker pool that potentially runs seed system or shoot control plane components should at least have **three nodes**. This is also the minium size that is required by an HA `etcd` cluster with a failure tolerance of a single node. Furthermore, the nodes must run in a single zone only (see [Recommended number of nodes and zones](#recommended-number-of-nodes-and-zones)).
+
+### Hosting a HA shoot control plane with `zone` failure tolerance
+
+To host an HA shoot control plane across availability zones, worker pools should have a minimum of **three nodes spread across an odd number of availability zones (min. 3)**.
+
+An additional label `seed.gardener.cloud/multi-zonal: true` should be added to the seed indicating that this seed is capable of hosting multi-zonal HA shoot control planes, which in turn will help gardener scheduler to short-list the seeds as candidates.
+
+In case of a `ManagedSeed` Gardener can add this label automatically to seed clusters if at least one worker pool fulfills the requirements mentioned above and doesn't enforce `Taints` on its nodes. Gardener may in addition validate if the `ManagedSeed` is properly set up for the `seed.gardener.cloud/multi-zonal: true` label when it is added manually.
+
+### Compute Seed Usage
+
+At present seed usage is computed by counting the number of shoot control planes that are hosted in a seed. Every seed has a number of shoots it can host `status.allocatable.shoots` (configurable via [ResourceConfiguration](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/gardenlet/apis/config/v1alpha1/types.go#L449). Operators need to rethink this value for multi-zonal seed clusters.
+
+Which parameters could be considered?
+* Number of available machines of a type as requested as part of the shoot spec. Sufficient capacity should be available to also allow rolling updates which will also be governed by `maxSurge` configuration at the worker pool level.
+* Node CIDR range must grant enough space to schedule additional replicas that the HA feature requires.
```

Comment:
Node CIDR is probably a bit misleading here. Expanding the node CIDR allows you to have more nodes, which are required in an HA scenario for sure, but if you discuss replicas it is most likely that you actually refer to the pod CIDR instead.
There is also the pod CIDR per node, which might be relevant. However, from my point of view it does not play an important role in the HA scenario.

Reply:
There are anti-affinity rules which mandates one replica per node. So the number of replicas has a direct bearing on the number of nodes that are required. To take an example: Today we have a single node etcd which requires a single node to be available. When we move to multi-node etcd with a default cluster size of 3 then 3 nodes will be requires at all times. This means the requirement of number of nodes for hosting a HA control plane will increase.

Reply:
The number of nodes for `etcd` will likely increase by factor 3 if and only if all shoot clusters of a seed will use the HA feature. However, this is not true necessarily true for the other components. For example, `gardener-resource-manager` is already scaled to 3 replicas.
Overall, it means that we need slightly more nodes. So yes, there is a need for a bit more node IPs, but realistically this should not be of a lot concern. The more important point is that the amount/ratio of nodes will changed compared to the status quo so that the sizing will be somewhat different.
I doubt that any reasonable operator has sized the node range so tight that this change would be a problem. Adding the subnets for the availability zones might be a bigger issue.

Reply:
>The number of nodes for etcd will likely increase by factor 3 if and only if all shoot clusters of a seed will use the HA feature.

This will eventually depend upon how many customers opt for HA control planes once this feature is GA. Since this has an impact on node CIDR we list this point here.

>The more important point is that the amount/ratio of nodes will changed compared to the status quo so that the sizing will be somewhat different.

Yes that is true.

> I doubt that any reasonable operator has sized the node range so tight that this change would be a problem

Ok. If that is the case then i would still like to keep this point here so that it is not forgotten as it still has an impact.
Also see https://github.tools.sap/kubernetes/landscape-stability/issues/42

> Adding the subnets for the availability zones might be a bigger issue.
Can you elaborate a bit more? I see that there are 3 subnets that are created per zone. At this point in time i am not completely aware of the purpose of each subnet.

Reply:
I guess my main point is that in the current form of the GEP you can specify by how much the node pool would change in the worst case, i.e. the node pool used for `etcd` will at maximum grow to 3 times its size. From my perspective, it is more helpful to be specific if you can be specific.
With regards to the subnets, you can take a look at the provider specific infrastructure configuration for aws and alicloud clusters. They require not only the general node CIDR, but also zone specific CIDRs. I have seen too many clusters where this was done in a way that later prevented extending the cluster to more than one zone. This is a real world problem if you think about making an existing seed cluster multi-zone aware.

Reply:
> the node pool used for etcd will at maximum grow to 3 times its size. From my perspective, it is more helpful to be specific if you can be specific.

Done.

>I have seen too many clusters where this was done in a way that later prevented extending the cluster to more than one zone.

I will add this point to the doc and this will be an input to @dguendisch when he plans to add additional zones to an existing seed which only has a single zone.